File: src/main/java/nu/peg/slack/pt/model/Response.java
package nu.peg.slack.pt.model;

import static nu.peg.slack.pt.model.Status.Error;
import static nu.peg.slack.pt.model.Status.Ok;

public class Response<T> {

    public Status status;
    public T data;
    public String error;

    public Response() {
    }

    public Response(T data) {
        this.status = Ok;
        this.data = data;
    }

    public Response(String error) {
        this.status = Error;
        this.error = error;
    }

    public Response(Status status, T data, String error) {
        this.status = status;
        this.data = data;
        this.error = error;
    }

    public static <E> Response<E> error(String error) {
        return new Response<>(Error, null, error);
    }

    public static <E> Response<E> ok(E data) {
        return new Response<>(Ok, data, null);
    }
}

File: src/main/kotlin/nu/peg/slack/pt/model/Response.kt
package nu.peg.slack.pt.model

import nu.peg.slack.pt.model.Status.Error
import nu.peg.slack.pt.model.Status.*

class Response<T> {

    var status: Status
    var data: T
    var error: String

    constructor() {}

    constructor(data: T) {
        this.status = Ok
        this.data = data
    }

    constructor(error: String) {
        this.status = Error
        this.error = error
    }

    constructor(status: Status, data: T, error: String) {
        this.status = status
        this.data = data
        this.error = error
    }

    companion object {

        fun <E> error(error: String): Response<E> {
            return Response<E>(Error, null, error)
        }

        fun <E> ok(data: E): Response<E> {
            return Response(Ok, data, null)
        }
    }
}
